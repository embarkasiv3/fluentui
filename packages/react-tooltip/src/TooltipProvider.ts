export * from './components/TooltipProvider/index';
