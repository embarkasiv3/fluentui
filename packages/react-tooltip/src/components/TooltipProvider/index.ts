export * from './TooltipProvider';
export * from './TooltipProvider.types';
export * from './renderTooltipProvider';
export * from './useTooltipContext';
export * from './useTooltipManager';
export * from './useTooltipProvider';
