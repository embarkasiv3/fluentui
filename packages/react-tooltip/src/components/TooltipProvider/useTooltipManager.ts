import * as React from 'react';
import { useFluent } from '@fluentui/react-shared-contexts';
import { TooltipImperativeHandle } from '../Tooltip/Tooltip.types';

/**
 * The tooltip manager is responsible for managing the visibiltiy of the tooltips,
 * including ensuring that only one tooltip is visible at once, and handling the
 * delay to show or hide a tooltip.
 *
 * This imperative interface is used by TooltipTrigger to show and hide its tooltip
 * based on events on the trigger element.
 *
 * {@docCategory TooltipProvider}
 */
export interface TooltipManager {
  /**
   * Show the given tooltip.
   * Call this in response to a PointerEnter or Focus event on the trigger.
   *
   * @param args - {@see ShowTooltipArgs}
   */
  showTooltip: (args: ShowTooltipArgs) => void;

  /**
   * Hide the tooltip triggered by the given element.
   * Call this in response to a PointerLeave or Blur event on the trigger.
   *
   * @param trigger - The element that lost focus
   */
  hideTooltip: (trigger: HTMLElement) => void;

  /**
   * Unconditionally hide any tooltip that's currently visible, without a delay.
   */
  hideAll: () => void;

  /**
   * Notify the manager that the pointer is inside the tooltip.
   */
  onPointerEnterTooltip: () => void;

  /**
   * Notify the manager that the pointer left the tooltip.
   */
  onPointerLeaveTooltip: () => void;
}

/**
 * The arguments to TooltipManager.showTooltip
 *
 * {@docCategory TooltipProvider}
 */
export type ShowTooltipArgs = {
  /**
   * The tooltip to show.
   */
  tooltip: TooltipImperativeHandle;

  /**
   * The element that triggered the tooltip.
   */
  trigger: HTMLElement;

  /**
   * The element for the tooltip to point to. If not defined, the trigger element will be used as the target.
   */
  target?: HTMLElement | null;

  /**
   * Delay before the tooltip is shown, in milliseconds
   *
   * @defaultvalue 250
   */
  showDelay?: number;

  /**
   * Delay before the tooltip is hidden, in milliseconds
   *
   * @defaultvalue 250
   */
  hideDelay?: number;
};

const defaultShowDelay = 250;
const defaultHideDelay = 250;

const useTimeout = () => {
  const [state] = React.useState(() => ({
    id: undefined as ReturnType<typeof setTimeout> | undefined,
    set: (fn: () => void, delay: number) => {
      state.clear();
      state.id = setTimeout(fn, delay);
    },
    clear: () => {
      if (state.id !== undefined) {
        clearTimeout(state.id);
        state.id = undefined;
      }
    },
  }));

  // Clean up the timeout when the component is unloaded
  React.useEffect(() => state.clear, [state]);

  return [state.set, state.clear] as const;
};

export const useTooltipManager = () => {
  const [setTooltipTimeout, clearTooltipTimeout] = useTimeout();

  const tooltipManager = React.useMemo<TooltipManager>(() => {
    // Keep track of the trigger element that is either hovered or focused
    let hovered: HTMLElement | undefined = undefined;

    // Keep track of the tooltip that is currently visible
    let visible: ShowTooltipArgs | undefined = undefined;

    const showTooltip = (args: ShowTooltipArgs) => {
      hovered = args.trigger;

      if (visible?.trigger === args.trigger) {
        // The tooltip is already visible; cancel the hide timer if it was running
        clearTooltipTimeout();
      } else {
        // Show the new tooltip immediately if there is already one visible
        const showDelay = visible ? 0 : args.showDelay;

        setTooltipTimeout(() => {
          // Double check that the trigger is still being hovered
          if (hovered === args.trigger) {
            hideAll();
            args.tooltip.show(args.target ?? args.trigger);
            visible = args;
          }
        }, showDelay ?? defaultShowDelay);
      }
    };

    const hideTooltip = (trigger: HTMLElement) => {
      // Ignore if this is coming from an element that's not the trigger
      if (hovered !== trigger) {
        return;
      }

      hovered = undefined;
      if (!visible) {
        // The tooltip is already hidden; cancel the show timer if it was running
        clearTooltipTimeout();
      } else {
        setTooltipTimeout(() => {
          // Double check that the trigger or tooltip didn't get hovered in the meantime
          if (!hovered || hovered !== visible?.trigger) {
            hideAll();
          }
        }, visible.hideDelay ?? defaultHideDelay);
      }
    };

    const hideAll = () => {
      clearTooltipTimeout();
      visible?.tooltip.hide();
      visible = undefined;
    };

    const onPointerEnterTooltip = () => {
      // Treat hovering over the tooltip the same as hovering over the tooltip's trigger element
      if (visible) {
        hovered = visible.trigger;
      }
    };

    const onPointerLeaveTooltip = () => {
      if (visible) {
        hideTooltip(visible.trigger);
      }
    };

    return { showTooltip, hideTooltip, hideAll, onPointerEnterTooltip, onPointerLeaveTooltip };
  }, [setTooltipTimeout, clearTooltipTimeout]);

  // Listen for the escape key on the document
  const { targetDocument } = useFluent();
  React.useEffect(() => {
    if (targetDocument) {
      const onKeyDown = (ev: KeyboardEvent) => {
        if (ev.key === 'Escape' || ev.key === 'Esc') {
          tooltipManager.hideAll();
        }
      };

      targetDocument.addEventListener('keydown', onKeyDown);
      return () => targetDocument.removeEventListener('keydown', onKeyDown);
    }
  }, [targetDocument, tooltipManager]);

  return tooltipManager;
};
