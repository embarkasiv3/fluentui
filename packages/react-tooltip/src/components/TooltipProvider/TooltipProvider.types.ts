import * as React from 'react';
import { ComponentProps, ComponentState } from '@fluentui/react-utilities';
import { TooltipManager } from './useTooltipManager';

/**
 * {@docCategory TooltipProvider}
 */
export interface TooltipProviderProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  // TooltipProvider has no additional props
}

/**
 * {@docCategory TooltipProvider}
 */
export type TooltipProviderState = ComponentState<
  React.RefObject<HTMLElement>,
  TooltipProviderProps & {
    tooltipManager: TooltipManager;
  }
>;
