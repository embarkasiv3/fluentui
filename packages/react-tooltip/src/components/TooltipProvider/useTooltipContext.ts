import * as React from 'react';
import { TooltipProps } from '../Tooltip/Tooltip.types';
import { TooltipManager } from './useTooltipManager';

/**
 * The context provided by TooltipProvider
 *
 * {@docCategory TooltipProvider}
 */
export type TooltipContext = {
  /**
   * The Tooltip component
   */
  TooltipComponent: React.FC<TooltipProps & React.RefAttributes<HTMLElement>>;

  /**
   * The instance of TooltipManager
   */
  tooltipManager: TooltipManager | undefined;

  /**
   * The root portal element for the tooltip
   */
  tooltipContainer: HTMLElement | undefined;
};

// eslint-disable-next-line @typescript-eslint/naming-convention
export const internal__TooltipContext = React.createContext<TooltipContext>({
  // These default values are replaced by TooltipProvider
  TooltipComponent: () => null, // eslint-disable-line @typescript-eslint/naming-convention
  tooltipManager: undefined,
  tooltipContainer: undefined,
});

export const useTooltipContext = () => React.useContext(internal__TooltipContext);
