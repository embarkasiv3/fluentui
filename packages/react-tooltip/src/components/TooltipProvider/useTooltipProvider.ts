import * as React from 'react';
import { makeMergeProps, useMergedRefs } from '@fluentui/react-utilities';
import { TooltipProviderProps, TooltipProviderState } from './TooltipProvider.types';
import { useTooltipManager } from './useTooltipManager';
import { useFluent } from '@fluentui/react-shared-contexts';

const mergeProps = makeMergeProps<TooltipProviderState>();

/**
 * Create the state required to render TooltipProvider.
 *
 * The returned state can be modified with hooks such as useTooltipProviderStyles,
 * before being passed to renderTooltipProvider.
 *
 * @param props - props from this instance of TooltipProvider
 * @param ref - reference to root HTMLElement of TooltipProvider
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory TooltipProvider}
 */
export const useTooltipProvider = (
  props: TooltipProviderProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: TooltipProviderProps,
): TooltipProviderState => {
  const { document } = useFluent();

  // createElement is not deprecated, but eslint seems to think it is
  // eslint-disable-next-line deprecation/deprecation
  const [tooltipContainer, setTooltipContainer] = React.useState(() => document?.createElement('div'));

  // If the document ever changes, need to re-create the tooltip container element
  if (tooltipContainer?.ownerDocument !== document) {
    // eslint-disable-next-line deprecation/deprecation
    setTooltipContainer(document?.createElement('div'));
  }

  const state = mergeProps(
    {
      ref: useMergedRefs(ref),
      tooltipContainer,
      tooltipManager: useTooltipManager(),
    },
    defaultProps,
    props,
  );

  React.useLayoutEffect(() => {
    const root = state.ref.current;
    if (root && tooltipContainer) {
      root.appendChild(tooltipContainer);
      return () => {
        root.removeChild(tooltipContainer);
      };
    }
  }, [state.ref, tooltipContainer]);

  return state;
};
