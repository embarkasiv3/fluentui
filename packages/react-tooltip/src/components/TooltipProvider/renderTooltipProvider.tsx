import * as React from 'react';
import { TooltipProviderState } from './TooltipProvider.types';
import { getSlots } from '@fluentui/react-utilities';
import { Tooltip } from '../../Tooltip';
import { internal__TooltipContext } from './useTooltipContext';

/**
 * Render the final JSX of TooltipProvider
 * {@docCategory TooltipProvider}
 */
export const renderTooltipProvider = (state: TooltipProviderState) => {
  const { slots, slotProps } = getSlots(state, []);

  const { children, ...rootProps } = slotProps.root;

  return (
    <internal__TooltipContext.Provider
      value={{
        tooltipManager: state.tooltipManager,
        TooltipComponent: Tooltip,
      }}
    >
      {children}
      <slots.root {...rootProps} />
    </internal__TooltipContext.Provider>
  );
};
