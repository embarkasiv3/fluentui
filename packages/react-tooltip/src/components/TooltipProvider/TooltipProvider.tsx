import * as React from 'react';
import { useTooltipProvider } from './useTooltipProvider';
import { TooltipProviderProps } from './TooltipProvider.types';
import { renderTooltipProvider } from './renderTooltipProvider';

/**
 * Provides the context for TooltipTriggers to render and display tooltips
 *
 * {@docCategory TooltipProvider}
 */
export const TooltipProvider = React.forwardRef<HTMLElement, TooltipProviderProps>((props, ref) => {
  const state = useTooltipProvider(props, ref);
  return renderTooltipProvider(state);
});

TooltipProvider.displayName = 'TooltipProvider';
