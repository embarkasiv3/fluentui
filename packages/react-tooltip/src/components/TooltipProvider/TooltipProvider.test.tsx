import * as React from 'react';
import { TooltipProvider } from './TooltipProvider';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('TooltipProvider', () => {
  isConformant({
    Component: TooltipProvider,
    displayName: 'TooltipProvider',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for TooltipProvider in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<TooltipProvider>Default TooltipProvider</TooltipProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
