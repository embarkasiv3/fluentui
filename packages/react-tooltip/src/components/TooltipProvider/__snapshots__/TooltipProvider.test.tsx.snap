// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipProvider renders a default state 1`] = `
Array [
  "Default TooltipProvider",
  <div />,
]
`;
